pkg/controller/job/backoff_utils.go

```diff
@@ -0,0 +1,213 @@
+		return u.key, nil
+	}
+	return "", fmt.Errorf("could not find key for obj %#v", obj)
+}
+
+func getNewBackoffInfo(backoffInfoStore *backoffInfoStore, key string, pods []*v1.Pod, newSucceededPods []*v1.Pod, newFailedPods []*v1.Pod) backoffInfo {
+	if backoffInfoStore.getBackoffInfo(key) == nil {
+		countedPods := getCountedPods(pods, newSucceededPods, newFailedPods)
+		failuresAfterLastSuccess, lastFailureTime := getInitialBackoffParameters(countedPods)
+		backoffInfoStore.updateBackoffInfo(key, failuresAfterLastSuccess, lastFailureTime)
+	}
+
+	backoffInf := backoffInfoStore.getBackoffInfo(key)
+	sortByFinishedTime(newSucceededPods)
+	sortByFinishedTime(newFailedPods)
+	newBackoffInfo := backoffInfo{
+		key,
+		backoffInf.failuresAfterLastSuccess,
+		backoffInf.lastFailureTime,
+	}
+
+	if len(newSucceededPods) == 0 && len(newFailedPods) == 0 {
+		return newBackoffInfo
+	}
+
+	if len(newFailedPods) == 0 {
+		newBackoffInfo.failuresAfterLastSuccess = 0
+		newBackoffInfo.lastFailureTime = nil
+		return newBackoffInfo
+	}
+
+	if len(newSucceededPods) == 0 {
+		newBackoffInfo.failuresAfterLastSuccess = newBackoffInfo.failuresAfterLastSuccess + int32(len(newFailedPods))
+		lastFailureTime := getFinishedTime(newFailedPods[len(newFailedPods)-1])
+		newBackoffInfo.lastFailureTime = &lastFailureTime
+		return newBackoffInfo
+	}
+
+	newBackoffInfo.failuresAfterLastSuccess = 0
+	newBackoffInfo.lastFailureTime = nil
+
+	lastSuccessTime := getFinishedTime(newSucceededPods[len(newSucceededPods)-1])
+	for i := len(newFailedPods) - 1; i >= 0; i-- {
+		failedPod := newFailedPods[i]
+		failedTime := getFinishedTime(failedPod)
+		if lastSuccessTime.After(failedTime.Time) || lastSuccessTime.Equal(&failedTime) {
+			break
+		}
+		if newBackoffInfo.lastFailureTime == nil {
+			newBackoffInfo.lastFailureTime = &failedTime
+		}
+		newBackoffInfo.failuresAfterLastSuccess = newBackoffInfo.failuresAfterLastSuccess + 1
+	}
+
+	return newBackoffInfo
+}
+
+func getInitialBackoffParameters(countedPods []*v1.Pod) (int32, *metav1.Time) {
+	sortByFinishedTime(countedPods)
+	failures := 0
+	var lastFailureTime *metav1.Time
+	for i := len(countedPods) - 1; i >= 0; i-- {
+		pod := countedPods[i]
+		if pod.Status.Phase == v1.PodSucceeded {
+			break
+		}
+
+		failures = failures + 1
+		if lastFailureTime == nil {
+			finishedTime := getFinishedTime(pod)
+			lastFailureTime = &finishedTime
+		}
+	}
+
+	return int32(failures), lastFailureTime
+}
+
+func sortByFinishedTime(pods []*v1.Pod) {
+	finishedTimeSortFunction := func(p1, p2 *v1.Pod) bool {
+		p1FinishTime := getFinishedTime(p1)
+		p2FinishTime := getFinishedTime(p2)
+
+		return p1FinishTime.Before(&p2FinishTime)
+	}
+
+	sort.Slice(pods, func(i, j int) bool {
+		return finishedTimeSortFunction(pods[i], pods[j])
+	})
+}
+
+func getFinishedTime(p *v1.Pod) metav1.Time {
+	if p.Status.Phase == v1.PodFailed || p.Status.Phase == v1.PodSucceeded {
+		finishTime := p.Status.ContainerStatuses[0].State.Terminated.FinishedAt
```

Comment:
IIUC we cannot be sure that `Terminated` is set. For example, an orphan pod, deleted by PodGC (see: https://github.com/kubernetes/kubernetes/blob/b7ad17978eaba508c560f81bab2fb9d0b256e469/pkg/controller/podgc/gc_controller.go#L342) may be in terminal phase by its containers will not be updated to `Terminated` as this is Kubelet's responsibility. I would thus protect it with `Terminated != nil`.

Reply:
What should we fallback to in case of `Terminated == nil`?

Reply:
Will podgc eventually set the Terminated? Or is it possible that `Terminated` is never set forever?

Reply:
In case of PodGc the `Terminated` is never set. To test that you may run a long running job, and delete the node `kubectl delete nodes/$nodeName`. 

Note that, we could provide a precise finish time in case of PodGC by using the `lastTransitionTime` for the `DisruptionTarget` condition, which is set in the same request, but it adds complexity.

I suggest, unless there are other opinions, to compute it as `max` for the containers in the `Terminated` state if all are in that state, otherwise apply the fallback to `now`.

Reply:
Done